api/services/bulk-docs.js

```diff
@@ -26,45 +19,96 @@ const checkForDuplicates = docs => {
   }
 };
 
-const generateBatches = (docs, batchSize) => {
+const generateBatches = (ids, batchSize) => {
   const batches = [];
-  while (docs.length > 0) {
-    const batch = docs.splice(0, batchSize);
+  while (ids.length > 0) {
+    const batch = ids.splice(0, batchSize);
     batches.push(batch);
   }
   return batches;
 };
 
-const generateBatchPromise = (batch, res, { isFinal } = {}) => {
-  return db.medic.bulkDocs(batch).then(result => {
-    let resString = JSON.stringify(result);
-    resString += isFinal ? '' : ',';
-    res.write(resString);
+const fetchDocs = ids => {
+  if (ids.length === 0) {
+    return Promise.resolve([]);
+  }
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(extractDocs);
+};
+
+const incrementAttemptCounts = (docs, attemptCounts) => {
+  docs.forEach(doc => {
+    attemptCounts[doc._id] = (attemptCounts[doc._id] || 0) + 1;
   });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
-      })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+const deleteDocs = (docsToDelete, deletionAttemptCounts, updateAttemptCounts, cumulativeResult = []) => {
+  const finishedModifications = cumulativeResult.map(docUpdate => docUpdate.id);
+  const parentMap = {};
+  let docsToUpdate;
+  let docsToModify;
+  docsToDelete.forEach(doc => doc._deleted = true);
+  incrementAttemptCounts(docsToDelete, deletionAttemptCounts);
+  return utils.updateParentContacts(docsToDelete, parentMap)
+    .then(updatedParents => {
+      docsToUpdate = updatedParents;
+      incrementAttemptCounts(docsToUpdate, updateAttemptCounts);
+      docsToModify = docsToDelete.concat(docsToUpdate).filter(doc => !finishedModifications.includes(doc._id));
+      return db.medic.bulkDocs(docsToModify);
+    })
+    .then(result => {
+      result = result.filter(docUpdate => docUpdate.status !== 404);
+      const deletionFailures = [];
+      const updateFailures = [];
+      const deletionIds = docsToDelete.map(doc => doc._id);
+      const updateIds = docsToUpdate.map(doc => doc._id);
+      const errorIds = result
+        .map((docUpdate, index) => docUpdate.error && (docUpdate.id || docsToModify[index]._id))
+        .filter(id => id);
+      errorIds.forEach(id => {
+        if (deletionIds.includes(id) && deletionAttemptCounts[id] < 4) {
+          deletionFailures.push(id);
+          result = result.filter(docUpdate => docUpdate.id !== id);
```

Comment:
I'm not sure if this matters for perf (it's probably going to be N^2 regardless), but maybe do this after you've pushed everything instead of every single time?

Reply:
Good point, it's been really difficult looking at this with fresh eyes cause I keep getting lost in all the old logic/restrictions and steps it took to get here